core/src/main/java/io/parsingdata/metal/expression/value/BinaryValueExpression.java

```diff
@@ -64,26 +65,33 @@ public BinaryValueExpression(final ValueExpression left, final ValueExpression r
     public abstract Optional<Value> eval(final Value leftValue, final Value rightValue, final ParseState parseState, final Encoding encoding);
 
     @Override
-    public ImmutableList<Optional<Value>> eval(final ParseState parseState, final Encoding encoding) {
+    public ImmutableList<Value> eval(final ParseState parseState, final Encoding encoding) {
         return evalLists(left.eval(parseState, encoding), right.eval(parseState, encoding), parseState, encoding);
     }
 
-    private ImmutableList<Optional<Value>> evalLists(final ImmutableList<Optional<Value>> leftValues, final ImmutableList<Optional<Value>> rightValues, final ParseState parseState, final Encoding encoding) {
+    private ImmutableList<Value> evalLists(final ImmutableList<Value> leftValues, final ImmutableList<Value> rightValues, final ParseState parseState, final Encoding encoding) {
         return reverse(padList(evalLists(leftValues, rightValues, parseState, encoding, new ImmutableList<>()).computeResult(), Math.abs(leftValues.size - rightValues.size)).computeResult());
     }
 
-    private Trampoline<ImmutableList<Optional<Value>>> evalLists(final ImmutableList<Optional<Value>> leftValues, final ImmutableList<Optional<Value>> rightValues, final ParseState parseState, final Encoding encoding, final ImmutableList<Optional<Value>> result) {
+    private Trampoline<ImmutableList<Value>> evalLists(final ImmutableList<Value> leftValues, final ImmutableList<Value> rightValues, final ParseState parseState, final Encoding encoding, final ImmutableList<Value> result) {
         if (leftValues.isEmpty() || rightValues.isEmpty()) {
             return complete(() -> result);
         }
-        return intermediate(() -> evalLists(leftValues.tail, rightValues.tail, parseState, encoding, result.add(leftValues.head.flatMap(leftValue -> rightValues.head.flatMap(rightValue -> eval(leftValue, rightValue, parseState, encoding))))));
+        return intermediate(() -> evalLists(leftValues.tail, rightValues.tail, parseState, encoding, result.add(checkEval(leftValues.head, rightValues.head, parseState, encoding))));
     }
 
-    private Trampoline<ImmutableList<Optional<Value>>> padList(final ImmutableList<Optional<Value>> list, final long size) {
+    private Trampoline<ImmutableList<Value>> padList(final ImmutableList<Value> list, final long size) {
         if (size <= 0) {
             return complete(() -> list);
         }
-        return intermediate(() -> padList(list.add(Optional.empty()), size - 1));
+        return intermediate(() -> padList(list.add(NOT_A_VALUE), size - 1));
+    }
+
+    private Value checkEval(final Value leftValue, final Value rightValue, final ParseState parseState, final Encoding encoding) {
```

Comment:
`check` seems a bit weird for the method name, though I can't think of a better one.

Reply:
`safeEval(...)` it is (see `UnaryValueExpression`-discussion).